### ISSUE TITLE
Supervisor fails to start Horizon after a server reboot

### ISSUE THREAD

**Issue by JackWH** [laravel/horizon#1130]
### Versions: 

- Horizon Version: 5.9.3
- Laravel Version: 9.4.1
- PHP Version: 8.1
- Redis Driver & Version: predis/phpredis 5.3.6
- Database Driver & Version: MariaDB 10.3

### Description:

I have a Forge server with the default Supervisor configuration enabled for Horizon. In the past few weeks I've noticed after the server is rebooted, Horizon fails to start — but the daemon still appears to be running.

Forge shows the daemon as being "active". But clicking on the daemon status returns this error:

```
daemon-632045:daemon-632045_00   FATAL     Exited too quickly (process log may have details)
```

Forge's daemon log shows the error below appearing 4 times in a row:

```
In Connection.php line 116:
                                                  
  LOADING Redis is loading the dataset in memory  
```

In Forge's control panel "Start Seconds" is set to 1, "Stop Seconds" is set to 10, and the "Stop Signal" is set to `SIGTERM`. As far as I'm aware these are the default settings.

My application logs seem to be catching this as a `RedisException`:

```
RedisException LOADING Redis is loading the dataset in memory 
    vendor/laravel/framework/src/Illuminate/Redis/Connections/Connection.php:116 Redis::get
    vendor/laravel/framework/src/Illuminate/Redis/Connections/Connection.php:116 Illuminate\Redis\Connections\Connection::command
    vendor/laravel/framework/src/Illuminate/Redis/Connections/PhpRedisConnection.php:532 Illuminate\Redis\Connections\PhpRedisConnection::command
    vendor/laravel/framework/src/Illuminate/Redis/Connections/PhpRedisConnection.php:55 Illuminate\Redis\Connections\PhpRedisConnection::get
    vendor/laravel/framework/src/Illuminate/Cache/RedisStore.php:62 Illuminate\Cache\RedisStore::get
    vendor/laravel/framework/src/Illuminate/Cache/Repository.php:97 Illuminate\Cache\Repository::get
    vendor/laravel/framework/src/Illuminate/Cache/CacheManager.php:418 Illuminate\Cache\CacheManager::__call
    vendor/laravel/telescope/src/Watchers/DumpWatcher.php:43 Laravel\Telescope\Watchers\DumpWatcher::register
    vendor/laravel/telescope/src/RegistersWatchers.php:48 Laravel\Telescope\Telescope::registerWatchers
    vendor/laravel/telescope/src/Telescope.php:141 Laravel\Telescope\Telescope::start
    vendor/laravel/telescope/src/TelescopeServiceProvider.php:33 Laravel\Telescope\TelescopeServiceProvider::boot
    vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php:36 Illuminate\Container\BoundMethod::Illuminate\Container\{closure}
    vendor/laravel/framework/src/Illuminate/Container/Util.php:41 Illuminate\Container\Util::unwrapIfClosure
    vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php:93 Illuminate\Container\BoundMethod::callBoundMethod
    vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php:37 Illuminate\Container\BoundMethod::call
    vendor/laravel/framework/src/Illuminate/Container/Container.php:653 Illuminate\Container\Container::call
    vendor/laravel/framework/src/Illuminate/Foundation/Application.php:928 Illuminate\Foundation\Application::bootProvider
    vendor/laravel/framework/src/Illuminate/Foundation/Application.php:909 Illuminate\Foundation\Application::Illuminate\Foundation\{closure}
    [internal] array_walk
    vendor/laravel/framework/src/Illuminate/Foundation/Application.php:910 Illuminate\Foundation\Application::boot
    vendor/laravel/framework/src/Illuminate/Foundation/Bootstrap/BootProviders.php:17 Illuminate\Foundation\Bootstrap\BootProviders::bootstrap
    vendor/laravel/framework/src/Illuminate/Foundation/Application.php:239 Illuminate\Foundation\Application::bootstrapWith
    vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php:310 Illuminate\Foundation\Console\Kernel::bootstrap
    vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php:127 Illuminate\Foundation\Console\Kernel::handle
    artisan:37 [main]
```

### Comments

I appreciate this might not strictly be a Horizon bug, but I didn't know where else to post this.

I would've thought Horizon would recognise and handle the `RedisException` in this case, and continue waiting for the dataset to finish loading into memory, rather than simply failing.

Is this something Horizon should handle when booting up? Or if not, any advice on how to resolve this? Currently I have to log in to Forge and restart the daemon each time.

### Steps To Reproduce:

1. Run Horizon on a Forge server
2. Reboot the server
3. Maybe get the error above?

**Comment by mfn** [comment]
What is the value of `startretries`? Even if Horizon doesn't come up, ideally supervisor would retry it a couple times.

**Comment by jbrooksuk** [comment]
Note that Forge is not showing you a **live** status of the Daemon here. Any changes to the status of the daemon made from outside of Forge (such as a restart, or manual daemon restarts) will not be reflected in the panel.

**Comment by driesvints** [comment]
@JackWH I'm afraid we cannot be of much help here. It seems to me that this is an issue specifically related to your daemon process rather than a Horizon issue. Can you please try a support channel?

- [Laracasts Forums](https://laracasts.com/discuss)
- [Laravel.io Forums](https://laravel.io/forum)
- [StackOverflow](https://stackoverflow.com/questions/tagged/laravel)
- [Discord](https://discordapp.com/invite/KxwQuKb)
- [Larachat](https://larachat.co)
- [IRC](https://web.libera.chat/?nick=laravelnewbie&channels=#laravel)

**Comment by JackWH** [comment]
> What is the value of `startretries`? Even if Horizon doesn't come up, ideally supervisor would retry it a couple times.

@mfn Thanks for responding. My Supervisor config is pasted below. FWIW I've never changed this, this is the default as Forge set it up. Looks like there's no `startretries` in here. The docs suggest the default is 3, would you advise upping this and seeing if it resolves?

```
[program:daemon-632045]
directory=/home/forge/my-app/deployments/current/
command=php artisan horizon

process_name=%(program_name)s_%(process_num)02d
autostart=true
autorestart=true
user=forge
numprocs=1
startsecs=1
redirect_stderr=true
stdout_logfile=/home/forge/.forge/daemon-632045.log
```

> Note that Forge is not showing you a **live** status of the Daemon here. Any changes to the status of the daemon made from outside of Forge (such as a restart, or manual daemon restarts) will not be reflected in the panel.

@jbrooksuk Thanks for the heads-up 👍

**Comment by mfn** [comment]
> would you advise upping this and seeing if it resolves?

When I look at the error message
> `LOADING Redis is loading the dataset in memory`

So _maybe_ (wild guess) your dataset is unusually big etc. so the combination of
- `startsecs=1` (1 = default)
- `startretries=3` (3 = default, as you figured)

It means within 3 seconds (probably roughly) it tries to start up horizon. If starting up redis due to the dataset takes longer, it won't work.

I guess you should measure how long redis takes and adjust those values.

Also there are tons of logs in `/var/log/supervisor/` likely providing you more insights. For cases where a process managed by supervisor does not come up, the `supervisord.log` in above directory is very useful.

(I coincidentally tuned a couple of systems in multiple environments and therefore invested a lot in understanding how this all plays together in supervisor; however I'm not using Horizon nor Forge so take it with a grain of salt 😬)

**Comment by JackWH** [comment]
@mfn Thanks for the tips... really helpful 👍 Looks like I've got it solved now.

You guessed correctly: I hadn't realised but Redis was storing ~2GB, which would explain why it's taking a few seconds.

Setting `startsecs=10` and `startretries=10` seemed to be the right combination to get it over the line. Unfortunately I'm still getting the `RedisException` logged each time, but that's not the end of the world. Maybe something for Horizon to handle differently in future.

I also deleted and re-created the daemon in Forge. I assume when I first set it up the default configuration was a little different — the newly-created daemon has these additional fields set in the config, which weren't present in the old one (just on the UI):

```
stopwaitsecs=10
stopsignal=SIGTERM
stopasgroup=true
killasgroup=true
```

Thanks all for your help 😊
